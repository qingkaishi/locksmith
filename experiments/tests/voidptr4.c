#include <pthread.h>

void *g; /* models a conflated container */
pthread_mutex_t l;

struct foo {
  int u;    /* no race */
};

struct bar {
  float v;  /* should be a race on this */
  int m;    /* but no race on this */
};

void *dostuff1(void *x) {
  struct foo *z;

  pthread_mutex_lock(&l);
  g = x;
  z = (struct foo *) g;
  pthread_mutex_unlock(&l);
  z->u = 5;
}

void *dostuff2(void *y) {
  struct bar *w;

  pthread_mutex_lock(&l);
  g = y;
  w = (struct bar *) g;
  pthread_mutex_unlock(&l);
  w->v = 5;
}

void *dostuff3(void *z) {
  struct bar *t;
  pthread_mutex_lock(&l);
  g = z;
  t = (struct bar *) g;
  pthread_mutex_unlock(&l);
  t->v = 6;
}

int main(int argc, char **argv) {
  struct foo a;
  struct bar b;
  pthread_t t1;
  pthread_mutex_init(&l, NULL);
  pthread_create(&t1, NULL, dostuff1, &a);
  pthread_create(&t1, NULL, dostuff2, &b);
  pthread_create(&t1, NULL, dostuff3, &b);
}
