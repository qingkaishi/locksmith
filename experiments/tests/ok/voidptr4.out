************************* STARTING *************************

shared: &g:voidptr4.c:3
        alloc.v:voidptr4.c:46
&g:voidptr4.c:3 is protected by:
  concrete l:voidptr4.c:48
  l:voidptr4.c:4

Warning: Possible data race: alloc.v:voidptr4.c:46 is not protected!
 references:
  dereference of *w.v:voidptr4.c:26 at voidptr4.c:32
    alloc.v:voidptr4.c:46 => b.v:voidptr4.c:46
                          => *w.v:voidptr4.c:26 at dostuff3 voidptr4.c:51
  locks acquired:
    <empty>
  in: main at voidptr4.c:44 

  dereference of *w.v:voidptr4.c:26 at voidptr4.c:41
    alloc.v:voidptr4.c:46 => b.v:voidptr4.c:46
                          => *w.v:voidptr4.c:26 at dostuff3 voidptr4.c:51
  locks acquired:
    <empty>
  in: main at voidptr4.c:44 


*************************** DONE ***************************
